lib-gphotos/albums.go

```diff
@@ -63,7 +63,9 @@ func (c *Client) AlbumByName(name string) (album *photoslibrary.Album, found boo
 // If the album doesn't exists it will try to create it.
 func (c *Client) GetOrCreateAlbumByName(name string) (*photoslibrary.Album, error) {
 	ctx := context.TODO()                    // TODO: ctx should be received (breaking change)
+	c.mu.RLock()                             // Prevent multiple Album search when we want to create an Album
 	album, found, err := c.AlbumByName(name) // TODO: ctx should be passed (breaking change)
+	c.mu.RUnlock()
```

Comment:
You're going to have to forgive me if this is a dumb question :)

Looking at the diff it looks like the mutex gets applied when checking the album but gets lifted before actually going ahead and creating it, which is again locked by a mutex.

I interpret it like this, but would need to test to be sure...

- Worker 1 acquires lock at line 66, continues to check if album exists
- Worker 2 gets to line 66 and waits for mutex unlock
- Worker 1 finishes checking for album exists (it doesn't), but then continues and unlocks whilst resuming execution on
- Worker 2 is then released and starts to check if the album exists - there is still a window of opportunity for this worker to complete the album exists check before worker 1 gets to line 79 and starts creating the album.

Or is it sufficient that when worker 1 locks again at line 77 this would block worker 2 from continuing?

Is it possible to retain the lock from line 66, deferring the unlock and effectively keep the whole of the func as a mutex?

Reply:
You're absolutely right... Album look up can take a lot of time, I was trying to avoid the rest of go routines  waits for this look up. But with the previous implementation, the function was not thread safe. 

Eventually, an album cache would be used to make the look up more performant.

Thanks @mjhatcher